File: pubnub-gson/pubnub-gson-impl/src/integrationTest/java/com/pubnub/api/integration/PNConfigurationIntegrationTests.java
package com.pubnub.api.integration;

import com.pubnub.api.PubNubException;
import com.pubnub.api.UserId;
import com.pubnub.api.integration.util.CustomLoggerTestImpl;
import com.pubnub.api.integration.util.ITTestConfig;
import com.pubnub.api.integration.util.Utils;
import com.pubnub.api.java.PubNub;
import com.pubnub.api.java.v2.PNConfiguration;
import com.pubnub.api.java.v2.PNConfigurationOverride;
import org.aeonbits.owner.ConfigFactory;
import org.junit.Assert;
import org.junit.Test;

import java.util.Arrays;

public class PNConfigurationIntegrationTests {
    final ITTestConfig itTestConfig = ConfigFactory.create(ITTestConfig.class, System.getenv());
    
    @Test
    public void createPubnubWithImmutableConfiguration() throws PubNubException {
        String expectedUuid = PubNub.generateUUID();

        PNConfiguration.Builder configBuilder = PNConfiguration.builder(new UserId(expectedUuid), itTestConfig.subscribeKey());
        PubNub pubNub = PubNub.create(configBuilder.build());

        Assert.assertEquals(expectedUuid, pubNub.getConfiguration().getUserId().getValue());
        Assert.assertEquals(itTestConfig.subscribeKey(), pubNub.getConfiguration().getSubscribeKey());

        configBuilder.setUserId(new UserId("unexpected"));
        configBuilder.build();

        Assert.assertEquals(expectedUuid, pubNub.getConfiguration().getUserId().getValue());
    }

    @Test
    public void createConfigurationWithConfigurationActionBlock() throws PubNubException {
        String expectedUuid = PubNub.generateUUID();

        PNConfiguration.Builder configBuilder = PNConfiguration.builder(new UserId(expectedUuid), itTestConfig.subscribeKey())
                .publishKey(itTestConfig.publishKey())
                .customLoggers( Arrays.asList(new CustomLoggerTestImpl()))
                .secretKey("fa")
                ;
        PubNub pubNub = PubNub.create(configBuilder.build());

        Assert.assertEquals(expectedUuid, pubNub.getConfiguration().getUserId().getValue());
        Assert.assertEquals(itTestConfig.subscribeKey(), pubNub.getConfiguration().getSubscribeKey());
        Assert.assertEquals(itTestConfig.publishKey(), pubNub.getConfiguration().getPublishKey());
        Assert.assertTrue(pubNub.getConfiguration().getCustomLoggers().get(0) instanceof CustomLoggerTestImpl);
    }

    @Test
    public void createConfigurationOverride() throws PubNubException {
        String expectedUuid = PubNub.generateUUID();

        PNConfiguration.Builder configBuilder = PNConfiguration.builder(new UserId(expectedUuid), itTestConfig.subscribeKey())
            .publishKey(itTestConfig.publishKey());

        PNConfiguration config = configBuilder.build();

        Assert.assertEquals(itTestConfig.subscribeKey(), config.getSubscribeKey());
        Assert.assertEquals(itTestConfig.publishKey(), config.getPublishKey());


        PNConfiguration overrideConfig = PNConfigurationOverride.from(config).publishKey("overridePublishKey").build();

        Assert.assertEquals(itTestConfig.subscribeKey(), overrideConfig.getSubscribeKey());
        Assert.assertEquals("overridePublishKey", overrideConfig.getPublishKey());
    }

    @Test
    public void useConfigurationOverrideWithPublish() throws PubNubException {
        String expectedUuid = PubNub.generateUUID();

        PNConfiguration.Builder configBuilder = PNConfiguration.builder(new UserId(expectedUuid), itTestConfig.subscribeKey())
                .publishKey("rubbishKey");
        PNConfiguration config = configBuilder.build();
        PubNub pubnub = PubNub.create(config);

        PNConfiguration overrideConfig = PNConfigurationOverride.from(config).publishKey(itTestConfig.publishKey()).build();

        pubnub.publish().channel(Utils.randomChannel()).message("message").overrideConfiguration(overrideConfig).sync();
        // no exception expected
    }
}

File: pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PNConfigurationIntegrationTests.kt
package com.pubnub.api.integration

import com.pubnub.api.PubNub
import com.pubnub.api.UserId
import com.pubnub.api.enums.PNLogVerbosity
import com.pubnub.api.integration.LoggingIntegrationTest.CustomLoggerTestImpl
import com.pubnub.api.v2.PNConfiguration
import com.pubnub.api.v2.PNConfigurationOverride
import com.pubnub.test.CommonUtils.randomChannel
import com.pubnub.test.Keys
import org.junit.Assert
import org.junit.Test

class PNConfigurationIntegrationTests : BaseIntegrationTest() {
    @Test
    fun `create pubnub with immutable configuration`() {
        val expectedUuid = PubNub.generateUUID()

        val configBuilder = PNConfiguration.builder(UserId(expectedUuid), Keys.subKey)
        val pubNub = PubNub.create(configBuilder.build())

        Assert.assertEquals(expectedUuid, pubNub.configuration.userId.value)
        Assert.assertEquals(Keys.subKey, pubNub.configuration.subscribeKey)

        configBuilder.userId = UserId("unexpected")
        configBuilder.build()

        Assert.assertEquals(expectedUuid, pubNub.configuration.userId.value)
    }

    @Test
    fun `create configuration with configuration action block`() {
        val expectedUuid = PubNub.generateUUID()
        val expectedAuthToken = "token"

        val configBuilder = PNConfiguration.builder(UserId(expectedUuid), Keys.subKey) {
            publishKey = Keys.pubKey
            authToken = expectedAuthToken
            logVerbosity = PNLogVerbosity.NONE
            customLoggers = listOf(CustomLoggerTestImpl())
        }
        val pubNub = PubNub.create(configBuilder.build())

        Assert.assertEquals(expectedUuid, pubNub.configuration.userId.value)
        Assert.assertEquals(Keys.subKey, pubNub.configuration.subscribeKey)
        Assert.assertEquals(Keys.pubKey, pubNub.configuration.publishKey)
        Assert.assertEquals(expectedAuthToken, pubNub.configuration.authToken)
        Assert.assertEquals(expectedAuthToken, pubNub.getToken())
    }

    @Test
    fun `create configuration override`() {
        val expectedUuid = PubNub.generateUUID()
        val expectedAuthToken = "token"

        val configBuilder = PNConfiguration.builder(UserId(expectedUuid), Keys.subKey) {
            publishKey = Keys.pubKey
            authToken = "token$expectedAuthToken"
        }
        val config = configBuilder.build()

        Assert.assertEquals(Keys.subKey, config.subscribeKey)
        Assert.assertEquals(Keys.pubKey, config.publishKey)

        val overrideConfig = PNConfigurationOverride.from(config).apply {
            publishKey = "overridePublishKey"
            authToken = expectedAuthToken
        }.build()

        Assert.assertEquals(Keys.subKey, overrideConfig.subscribeKey)
        Assert.assertEquals("overridePublishKey", overrideConfig.publishKey)
        Assert.assertEquals(expectedAuthToken, overrideConfig.authToken)
    }

    @Test
    fun `use configuration override with Publish`() {
        val expectedUuid = PubNub.generateUUID()
        val expectedAuthToken = "token"

        val configBuilder = PNConfiguration.builder(UserId(expectedUuid), Keys.subKey) {
            publishKey = "rubbishKey"
            authToken = "old$expectedAuthToken"
        }
        val config = configBuilder.build()
        val pubnub = PubNub.create(config)

        val overrideConfig = PNConfigurationOverride.from(config).apply {
            publishKey = Keys.pubKey
            authToken = expectedAuthToken
        }.build()

        pubnub.publish(randomChannel(), "message").overrideConfiguration(overrideConfig).sync()
        // no exception expected
    }

    @Test
    fun `use configuration override builder with Publish`() {
        val expectedUuid = PubNub.generateUUID()
        val expectedAuthToken = "token"

        val configBuilder = PNConfiguration.builder(UserId(expectedUuid), Keys.subKey) {
            publishKey = "rubbishKey"
            authToken = "old$expectedAuthToken"
        }
        val pubnub = PubNub.create(configBuilder.build())

        pubnub.publish(randomChannel(), "message").overrideConfiguration {
            publishKey = Keys.pubKey
            authToken = expectedAuthToken
        }.sync()
        // no exception expected
    }
}
